doc/floppymeter.1
.TH floppymeter 1 "03Mar05" fdutils-5.5
.SH Name
floppymeter - measure raw capacity and exact rotation speed of floppy drive
'\" t
.de TQ
.br
.ns
.TP \\$1
..

.tr \(is'
.tr \(if`
.tr \(pd"

.SH Note
This manpage has been automatically generated from fdutils's texinfo
documentation.  However, this process is only approximative, and some
items, such as cross-references, footnotes and indices are lost in this
translation process.  Indeed, these items have no appropriate
representation in the manpage format.  Moreover, only the items specific
to each command have been translated, and the general information about
fdutils has been dropped in the manpage version.  Thus I strongly advise
you to use the original texinfo doc.
.TP
* \ \ 
To generate a printable copy from the texinfo doc, run the following
commands:
 
.nf
.ft 3
.in +0.3i
    ./configure; make dvi; dvips fdutils.dvi
.fi
.in -0.3i
.ft R
.lp
 
\&\fR
.TP
* \ \ 
To generate a HTML copy,  run:
 
.nf
.ft 3
.in +0.3i
    ./configure; make html
.fi
.in -0.3i
.ft R
.lp
 
\&\fRA pre-made HTML can be found at:
\&\fR\&\f(CW\(ifhttp://www.tux.org/pub/knaff/fdutils\(is\fR
.TP
* \ \ 
To generate an info copy (browsable using emacs' info mode), run:
 
.nf
.ft 3
.in +0.3i
    ./configure; make info
.fi
.in -0.3i
.ft R
.lp
 
\&\fR
.PP
The texinfo doc looks most pretty when printed or as HTML.  Indeed, in
the info version certain examples are difficult to read due to the
quoting conventions used in info.
.SH Description
.iX "p floppymeter"
.iX "c rotation speed"
.iX "c raw capacity"
.PP
 
.nf
.ft 3
.in +0.3i
\&\fR\&\f(CWfloppymeter [\fR\&\f(CW-f] [\fR\&\f(CW-w \fIwarmup-delay\fR\&\f(CW] [\fR\&\f(CW-W \fIwindow\fR\&\f(CW]
[\fR\&\f(CW-c \fIcycles\fR\&\f(CW] [\fR\&\f(CW-h] \fIdrive\fR\&\f(CW
.fi
.in -0.3i
.ft R
.lp
 
\&\fR
.PP
The \fR\&\f(CWfloppymeter\fR program measures characteristic parameters of a
floppy drive and floppy controller, such as the rotation speed of the
drive, the data transfer rate of the controller, and the resulting raw
capacity of a disk track.  To use this program, insert a disposable
floppy in the drive, and type \fR\&\f(CWfloppymeter --\fR\fIdensity\fR, where
density describes the density of the disk used for the test. (Should be
one of \fR\&\f(CWdd, hd\fR or \fR\&\f(CWed\fR). \fBCAUTION: the data on the
disk will be erased\fR.  This program should be used to verify whether the
drive and controller are out of tolerance if you experience problems
with some high capacity formats.  It only needs to be run once per
drive: although a disk is needed to perform the measurements, the
measured data only depend on the drive and the controller, and not on
the disk.
.PP
To measure the raw capacity of the disk track, the floppymeter program
formats the first track of the drive in a special way that allows it to
read the raw data (gaps and headers) of the disk. \fBThus, all data
previously stored on that disk is lost\fR.
.PP
The rotation speed is measured by timing the return time of a
\&\fR\&\f(CWreadid\fR command.  In order to gain more precision, the command is
issued many times in a row. During this phase, the number of rotations
since the start of the test, the average time per rotation since the start,
and a sliding average of the times of the last 30 rotations is printed,
and updated continuously.
.PP
The data transfer rate is deduced from the two parameters above.
.PP
At the end of the program, all parameters (raw capacity, duration of one
rotation, and data transfer rate) are printed again, as well as their
relative deviation to the standard value.  Finally, it suggests a
capacity deviation description line, which can be directly pasted into
the drive definition file (See section Drive descriptions.).
.PP
Usually, the data transfer rate should not deviate more than 150 ppm
from the expected value, and the rotation speed of the drive should not
deviate more than 3000 ppm from the expected value.  If these deviations
are bigger, you may get problems with certain high capacity formats.
.PP
If the raw capacity of the drive is too small, some high capacity
formats may become unformattable on this drive/controller combo.
.PP
If on the other hand, the raw capacity of the drive is too big, you may
get problems when writing to a disk formatted by this drive on another
drive with a smaller raw capacity. In order to avoid this, increase
superformats gap parameter (\fR\&\f(CW-G\fR).
.IP
.TP
\&\fR\&\f(CW-h\fR\ 
Prints a short help
.TP
\&\fR\&\f(CW--dd\fR\ 
Tells the program that we use a Double Density disk.
.TP
\&\fR\&\f(CW--hd\fR\ 
Tells the program that we use a High Density disk.
.TP
\&\fR\&\f(CW--ed\fR\ 
Tells the program that we use an Extra Density disk.
.TP
\&\fR\&\f(CW-f\fR\ 
Runs the measurement non interactively.  With this option, the program
doesn't ask for confirmation, and doesn't display the continuously
updated values during the rotation speed measurement.
.TP
\&\fR\&\f(CW-W\ \fIWindow\fR\&\f(CW\fR\ 
This value describes how many rotations are used for the computation of
the sliding average. Default is 30.
.TP
\&\fR\&\f(CW-c\ \fIcycles\fR\&\f(CW\fR\ 
Describes the number of rotations clocked during the rotations speed
determination test. Default is 1000.
.PP
.SH Bugs
.PP
This program is quite new, and may have bugs. Here are a few suggested
tests to check its sanity:
.TP
* \ \ 
The deviation of the data transfer rate solely depends on the
controller. It should not be different between two drives connected to
the same controller. However, the drive rotation speed may be different
for different drives.
.TP
* \ \ 
All data transfer rates (for double, high and extra density) are derived
from a same master frequency. Thus the \fIdeviation\fR of the data
transfer rate should be independent of the density used.
.SH See Also
Fdutils' texinfo doc


Changelog
020203	Fix fdmount and superformat bugs
020203	Create /dev/fd0 before testing for existence of /dev/fd0
011002	Fix floppycontrol.c printf conditional on FD_DISK_CHANGED, so that
	it works with gcc3 which apparently considers printf as a macro...
000612	Explicitly include sysmacros.h, apparently this is needed in
	certain environments
000603	Released 5.4
000527	Fixed a couple of buffer overrun exploits and other cleanup. Made
	fdutils compilable with Linux 2.3/2.4 kernels. Documentation fixes
970104	Fixed a few bugs listed in the Debian buglist posted to
	linux.debian.bugs.dis
	1.  Lessen dependancy of superformat on mtools by not failing if
	mformat cannot be called: low-level verify will still be done, but
	mbadblocks will be skipped
	2.  Fixed a few instances where the return value of getopt was
	misdeclared as a char instead of an integer
	3.  Fixed device names in fd.4
970411	Added error check to capacity measurement
	Corrected documentation Makefile bug
970629	Added vfat support to fdmount
970609	Fixed bug where a skew was erroneously applied to a 2m track
	Fixed typo in printfdprm
970604	Added vfat option for fdmount
970516	5.2 Released
970516	Fixed sunsite location in doc.
970430	Fixed typoes in doc about autodetection.
970428	Fixed manpage installation
970427	5.1 Released.  These repeated complaints about yylineno not
	defined start getting annoying ;-)
970426	More Makefile fixes.  Added note to doc about Alpha site
970422	Various Makefile fixes
970421	Lex yylineno fixes
970419	5.0 Released


src/floppymeter.c
#include <config.h>
#include <sys/types.h>
#ifdef HAVE_SYS_SYSMACROS_H
# include <sys/sysmacros.h>
#endif
#include <unistd.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <linux/fd.h>
#include <linux/fdreg.h>
#include <sys/ioctl.h>
#include <sys/time.h>
#include <fcntl.h>
#include <sys/stat.h>
#include <linux/major.h>
#include <getopt.h>
#include "enh_options.h"
#include "fdutils.h"
#include "driveprm.h"

#define NON_INTERACTIVE 1

long long gettime(void)
{
  struct timeval tv;
  int ret;

  ret = gettimeofday(&tv, 0);
  if ( ret < 0 ){
    perror("get time of day\n");
    exit(1);
  }

  return ( (long long) tv.tv_usec + ( (long long) tv.tv_sec ) * 1000000LL);
}

int dtr[4] = { 500000, 300000, 250000, 1000000 };

struct {
	int time_per_rot; /* nominal time per rotation */
	int sp_offset;
	int rate[3];
} cmos_table[]= {
  {      0, 0, { 0x100, 0x100, 0x100} }, /* no drive */
  { 200000, 0, { 0x002, 0x100, 0x100} }, /* 5 1/4 DD */
  { 200000, 0, { 0x002, 0x100, 0x100} }, /* 3 1/2 DD */
  { 166666, 0, { 0x001, 0x000, 0x100} }, /* 5 1/4 HD */
  { 200000, 0, { 0x002, 0x000, 0x100} }, /* 3 1/2 HD */
  { 200000, 0, { 0x002, 0x000, 0x043} }, /* 3 1/2 ED */
  { 200000, 0, { 0x002, 0x000, 0x043} }  /* 3 1/2 ED */
};


struct reg {
	long long time;
	int rot;
};


int sliding_avg(struct reg *reg, int n)
{
	int i;
	long long sum_x, sum_y, sum_xy, sum_x2, sum2_x;

	sum_x = sum_y = sum_xy = sum_x2 = 0;
	for(i=0; i<n; i++) {
		sum_x += reg[i].rot;
		sum_y += reg[i].time;
		sum_xy += reg[i].rot * reg[i].time;
		sum_x2 += reg[i].rot * reg[i].rot;
	}	
	sum2_x = sum_x * sum_x;
	return (n*sum_xy - sum_x * sum_y) / (n*sum_x2 - sum2_x);
}

int main(int argc, char **argv)
{
	int n;
	long long shall_cap;
	int fd=-2;
	long long sum_x, sum_y, sum_xy, sum_x2, sum2_x;
	long long avg, base_time, last_time, time, shall_time, my_dtr;
	struct floppy_drive_params dpr;
	char c;
	int ch;
	int rot;
	int missed;
	int j;
	int dn,rate, cap;
	struct stat buf;
	char *name;
	int cycles=1000;
	int warmup=100;
	int mask=0;
	int density=3;
	int window=10;
	int cylinder=0;

	struct enh_options optable[] = {
		{ 'C', "cycles", 1, EO_TYPE_LONG , 0, 0, 
		  (void *) &cycles,
		  "run during n cycles" },

		
		{ 'f', "force", 0, EO_TYPE_NONE, 0, NON_INTERACTIVE, 0,
		  "don't ask for confirmation" },

		{ 'w', "warmup", 1, EO_TYPE_LONG, 0, 0, (void *) &warmup,
		  "number of warm-up rotations" },

		{ 'W', "window", 1, EO_TYPE_LONG, 0, 0, (void *) &window,
		  "window for sliding average" },

		{ 'c', "cylinder", 1, EO_TYPE_LONG, 0, 0, (void *) &cylinder,
		  "cylinder used for test" },


		{ '\0', "dd", 0, EO_TYPE_LONG, 0, 0, (void *) &density,
		  "perform test on a double density disk" },

		{ '\0', "hd", 0, EO_TYPE_LONG, 1, 0, (void *) &density,
		  "perform test on a high density disk" },

		{ '\0', "ed", 0, EO_TYPE_LONG, 2, 0, (void *) &density,
		  "perform test on an extra density disk" },

		{ '\0', 0 }
	};

	struct reg *reg;

	while((ch=getopt_enh(argc, argv, optable, 
			     0, &mask, "[drive]") ) != EOF ){
		if ( ch== '?' ){
			fprintf(stderr,"exiting\n");
			exit(1);
		}
		printf("unhandled option %d\n", ch);
		exit(1);
	}


	if ( optind < argc )
		name = argv[optind];
	else
		name = "/dev/fd0";
	fd = open(name, 3 | O_NDELAY);
	if ( fd < 0 ){
		perror("can't open floppy drive");
		print_usage_exit(argv[0],optable,"[drive]");
	}

	if(!(mask & NON_INTERACTIVE)) {
		fprintf(stderr,
			"Warning: all data on the floppy disk will be lost. Continue (y/N)? ");
		c=getchar();
		fprintf(stderr, "\n");
		if(c != 'y' && c != 'Y')
			exit(1);
	}

	if (fstat (fd, &buf) < 0) {
		perror("fstat");
		exit(1);
	}
	if (!S_ISBLK(buf.st_mode) || major(buf.st_rdev) != FLOPPY_MAJOR) {
		fprintf(stderr,"%s is not a floppy drive\n", name);
		exit(1);
	}
	dn = minor( buf.st_rdev );
	dn = (dn & 3) + ((dn & 0x80) >> 5);

	if(ioctl(fd, FDGETDRVPRM, &dpr) < 0) {
		perror("get drive parameter");
		exit(1);
	}

	if(dpr.cmos < 1 || dpr.cmos > 6) {
		fprintf(stderr, "Bad cmos type\n");
		exit(1);
	}

	if(density == 3) {
		do {
			density--;
			rate = cmos_table[(int) dpr.cmos].rate[density];
		} while(rate >= 256);
	} else
		rate = cmos_table[(int) dpr.cmos].rate[density];
	if(rate >= 256) {
		fprintf(stderr,"density not supported by this drive\n");
		exit(1);
	}

	/* measure its capacity */
	measure_raw_capacity(fd, dn, rate, 
			     cylinder, warmup, !(mask & NON_INTERACTIVE));

	readid(fd, dn, rate, cylinder);
	base_time = last_time = gettime();
	missed = rot = 0;
	avg = cmos_table[(int)dpr.cmos].time_per_rot;
	reg = (struct reg *) calloc(window, sizeof(struct reg));
	for(j=0; j<window; j++) {
		reg[j].rot = 0;
		reg[j].time = base_time;
	}
	j=0;
	n=1;
	if(!(mask & NON_INTERACTIVE)) {
		fprintf(stderr,
			" _____________________________________________\n");
		fprintf(stderr,
			"|rotations  |average    |sliding    |missed   |\n");
		fprintf(stderr,
			"|since start|time per   |average of |rotations|\n");
		fprintf(stderr,
			"|of test    |rotation   |the last   |         |\n");
		fprintf(stderr,
			"|           |since start|%-8d   |         |\n",window);
		fprintf(stderr,
			"|           |of test    |rotations  |         |\n");
		fprintf(stderr,
			"|===========|===========|===========|=========|\n");
	}

	sum_x = sum_y = sum_xy = sum_x2 = 0;
	sum_y = base_time;
	while(rot < cycles) {
		readid(fd, dn, rate, cylinder);
		rot++;
		time = gettime();
		if(time - last_time > avg * 3 / 2) {
			/* if it takes unusually long, we may have missed one rotation */
			rot += (time - last_time - avg / 2) / avg;
			missed += (time - last_time - avg / 2) / avg;
		}

		last_time = time;
		reg[j].time = time;
		reg[j].rot = rot;

		sum_x += rot;
		sum_y += time;
		sum_xy += rot * time;
		sum_x2 += rot * rot;
		sum2_x = sum_x * sum_x;
		n++;

		avg = ((n*sum_xy - sum_x*sum_y) / (n*sum_x2 - sum2_x));
		if(!(mask & NON_INTERACTIVE)){
			fprintf(stderr,"|%6d     |%8d   |%8d   |%3d      |\r", 
				rot, (int)avg, sliding_avg(reg,window), missed);
			fflush(stderr);
		}
		j++;
		if(j>=window)
			j=0;
	}	
	if(!(mask & NON_INTERACTIVE)) {
		fprintf(stderr,"\n");
		fprintf(stderr,
			"|___________|___________|___________|_________|\n");
	}

	cap = measure_raw_capacity(fd, dn, rate, 
				   cylinder, 0, !(mask & NON_INTERACTIVE));

	shall_time = cmos_table[(int)dpr.cmos].time_per_rot;
	shall_cap = dtr[rate] * 2 * shall_time / 1000000LL;

	printf("\n");
	printf("capacity=%d half bits (should be %d half bits)\n", cap,
	       (int) shall_cap);
	printf("time_per_rotation=%d microseconds (should be %d)\n", (int) avg,
	       (int) shall_time);
	my_dtr = ((long long)cap) * 500000 / avg;
	printf("data transfer rate=%d bits per second (should be %d)\n", 
	       (int) my_dtr, dtr[rate]);
	printf("\n");
	printf("deviation on capacity: %+d ppm\n", 
	       (int) ((cap - shall_cap) * 1000000 / shall_cap));
	printf("deviation on time_per_rotation: %+d ppm\n",
	       (int) ((avg - shall_time) * 1000000 / shall_time));
	printf("deviation on data transfer rate: %+d ppm\n", 
	       (int) ((my_dtr - dtr[rate]) * 1000000 / dtr[rate]));

	printf("\nInsert the following line to your " DRIVEPRMFILE " file:\n");
	printf("drive%d: deviation=%d\n\n", 
	       dn, (int) ((cap - shall_cap) * 1000000 / shall_cap ));

#if 0
	margin = (shall_cap - cap + 15)/10;
	if(margin > 0)
		printf("recommended margin: %d\n", margin);
#endif
	exit(0);
}
